modules/packages/DistributedDeque.chpl

```diff
@@ -59,638 +57,709 @@ use Collection;
+  */
+  config param distributedDequeBlockSize = 8;
 
-        // Push...
-        tail.pushBack(_elt);
-        size.add(1);
+  /*
+    The ordering used for serial iteration. ``NONE``, the default, is the most performant
+    and is algorithmically similar to parallel iteration.
+  */
+  enum Ordering {
+    NONE,
+    FIFO,
+    LIFO
+  }
 
-        lock$;
+  pragma "no doc"
+  class DistributedDequeRC {
+    type eltType;
+    var _pid : int;
+
+    proc deinit() {
+      chpl_getPrivatizedCopy(DistributedDequeImpl(eltType), _pid).Destroy();
+      coforall loc in Locales do on loc {
+        delete chpl_getPrivatizedCopy(DistributedDequeImpl(eltType), _pid);
       }
+      
     }
   }
 
-  proc popBack() : eltType {
-    var elt : eltType;
-    on this {
-      var _elt : eltType;
-      local {
-        while true {
-          // Check if there is an element for us...
-          if size.read() == 0 {
-            while size.read() == 0 {
-              chpl_task_yield();
-            }
-          }
+  /*
+    A parallel-safe scalable distributed double-ended queue that supports both
+    insertion and removal from either end of the queue. Can be used as a Queue,
+    Stack, or even a List.
+  */
+  record DistDeque {
+    type eltType;
+    /*
+      The implementation of the Deque is forwarded. See :class:`DistributedDequeImpl` for
+      documentation.
+    */
+    var _impl : DistributedDequeImpl(eltType);
+
+    // Privatization id
+    pragma "no doc"
+    var _pid : int;
+    // Reference counting
+    pragma "no doc"
+    var _rc : Shared(DistributedDequeRC(eltType));
+
+    proc DistDeque(type eltType, cap = -1, targetLocales = Locales) {
+      _pid = (new DistributedDequeImpl(eltType, cap, targetLocales)).pid;
+      _rc = new Shared(new DistributedDequeRC(eltType, _pid = _pid));
+    }
 
-          lock$ = true;
+    inline proc _value {
+      if _pid == -1 {
+        halt("DistDeque is uninitialized...");
+      }
+      return chpl_getPrivatizedCopy(DistributedDequeImpl(eltType), _pid);
+    }
 
-          // Someone else came in and took a value, wait for the next one...
-          if size.read() == 0 {
-            lock$;
-            continue;
-          }
+    pragma "no doc"
+    inline proc these(param order : Ordering = Ordering.NONE) {
+      return _value.these(order);
+    }
 
-          // Pop...
-          _elt = tail.popBack();
-          if tail.isEmpty {
-            var node = tail;
-            tail = tail.prev;
-
-            // If not empty, remove reference to retired node...
-            // If it is empty, fix the head...
-            if tail != nil {
-              tail.next = nil;
-            } else {
-              head = nil;
-            }
+    pragma "no doc"
+    inline proc these(param order : Ordering = Ordering.NONE, param tag) where
+        (tag == iterKind.leader || tag == iterKind.standalone) &&
+        __primitive("method call resolves", _value, "these", tag=tag)

```

Comment:
You shouldn't need the "method call resolves" business here. And I don't remember seeing these forwarding  'these' in DistributedBag but maybe I missed them. Anyway, you know if you implemented leader/follower or standalone so you can just directly create the versions that you implemented, instead of using "method call resolves". That is a workaround appropriate for the internal array code.

Reply:
After removing and testing, it turns out they are needed, I get `unresolved call 'DistributedDequeImpl(int(64)).these(Ordering, tag=iterKind)'` otherwise.